src/repositories/updater.rs

```diff
@@ -0,0 +1,354 @@
+                    "no `{}` repositories stats needed to be updated",
+                    updater.host()
+                );
+                continue;
+            }
+            for chunk in needs_update.chunks(updater.chunk_size()) {
+                let res = updater.fetch_repositories(chunk)?;
+                for node in res.missing {
+                    Self::delete_repository(&mut conn, &node, updater.host())?;
+                }
+                for (_, repo) in res.present {
+                    Self::store_repository(&mut conn, updater.host(), repo)?;
+                }
+            }
+            info!("finished updating `{}` repositories stats", updater.host());
+        }
+        Ok(())
+    }
+
+    pub fn backfill_repositories(&self, ctx: &dyn Context) -> Result<()> {
+        let pool = ctx.pool()?;
+        let mut conn = pool.get()?;
+        for updater in &self.updaters {
+            info!(
+                "started backfilling `{}` repositories stats",
+                updater.host()
+            );
+
+            let needs_backfilling = conn.query(
+                "SELECT releases.id, crates.name, releases.version, releases.repository_url
+                 FROM releases
+                 INNER JOIN crates ON (crates.id = releases.crate_id)
+                 WHERE repository_id IS NULL AND repository_url LIKE $1;",
+                &[&format!("%{}%", updater.host())],
+            )?;
+
+            let mut missing_urls = HashSet::new();
+            for row in &needs_backfilling {
+                let id: i32 = row.get("id");
+                let name: String = row.get("name");
+                let version: String = row.get("version");
+                let url: String = row.get("repository_url");
+
+                if missing_urls.contains(&url) {
+                    debug!("{} {} points to a known missing repo", name, version);
+                } else if let Some(node_id) = self.load_repository_inner(&mut conn, &url)? {
+                    conn.execute(
+                        "UPDATE releases SET repository_id = $1 WHERE id = $2;",
+                        &[&node_id, &id],
+                    )?;
+                    info!(
+                        "backfilled `{}` repositories for {} {}",
+                        updater.host(),
+                        name,
+                        version,
+                    );
+                } else {
+                    debug!(
+                        "{} {} does not point to a {} repository",
+                        name,
+                        version,
+                        updater.host(),
+                    );
+                    missing_urls.insert(url);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn get_icon_name(&self, host: &str) -> &'static str {
+        for updater in &self.updaters {
+            if updater.host() == host {
+                return updater.icon();
+            }
+        }
+        ""
```

Comment:
I'd return a generic repository icon instead of an empty one.

Reply:
Good idea!